gapic-common/lib/gapic/rest/error.rb

```diff
@@ -37,36 +50,85 @@ class << self
+          status = nil
+          details = nil
+          headers = err.response_headers
 
           if err.response_body
-            msg, code = try_parse_from_body err.response_body
+            msg, code, status, details = try_parse_from_body err.response_body
             message = "An error has occurred when making a REST request: #{msg}" unless msg.nil?
             status_code = code unless code.nil?
           end
 
-          Gapic::Rest::Error.new message, status_code
+          Gapic::Rest::Error.new message, status_code, status: status, details: details, headers: headers
         end
 
         private
 
         ##
+        # @private
         # Tries to get the error information from the JSON bodies
         #
         # @param body_str [String]
         # @return [Array(String, String)]
         def try_parse_from_body body_str
           body = JSON.parse body_str
-          return [nil, nil] unless body && body["error"].is_a?(Hash)
 
-          message = body["error"]["message"]
-          code = body["error"]["code"]
+          unless body.is_a?(::Hash) && body&.key?("error") && body["error"].is_a?(::Hash)
+            return [nil, nil, nil,
+                    nil]
+          end
+          error = body["error"]
+
+          message = error["message"] if error.key? "message"
+          code = error["code"] if error.key? "code"
+          status = error["status"] if error.key? "status"
 
-          [message, code]
+          details = parse_details error["details"] if error.key? "details"
+
+          [message, code, status, details]
         rescue JSON::ParserError
-          [nil, nil]
+          [nil, nil, nil, nil]
+        end
+
+        ##
+        # @private
+        # Parses the details data, trying to extract the Protobuf.Any objects
+        # from it, if it's an array of hashes. Otherwise returns it as is.
+        #
+        # @param details [Object, nil] the details object
+        #
+        # @return [Object, nil]
+        def parse_details details
+          # For rest errors details will contain json representations of `Protobuf.Any`
+          # decoded into hashes. If it's not an array, of its elements are not hashes,
+          # it's some other case
+          return details unless details.is_a? ::Array
+
+          details.map do |detail_instance|
+            # Next, parse detail_instance into a Proto message.
+            # There are three possible issues for the JSON->Any->message parsing
+            # - json decoding fails
+            # - the json belongs to a proto message type we don't know about
+            # - any unpacking fails
+            # If we hit any of these three issues we'll just return the original hash
+            begin
+              any = ::Google::Protobuf::Any.decode_json detail_instance.to_json
+              klass = ::Google::Protobuf::DescriptorPool.generated_pool.lookup(any.type_name).msgclass
+              return detail_instance if klass.nil?
```

Comment:
I think you want `next detail_instance` which just returns `detail_instance` from the block (and hence passes the value through unchanged in the map operation) instead of `return detail_instance` which exits the entire method and returns only that one element. Same for two lines down.

(Maybe warrants a test, too?)

Reply:
:+1: